search/median_search2.cpp

```diff
@@ -0,0 +1,139 @@
+ * twice as fast as the slow pointer, when the fast pointer reaches the end of
+ * the list the slow pointer will be pointing to the middle node of the list.
+ *
+ * Here are some example lists you can use to see how the algorithm works
+ * A = [1,2,3,4,5]
+ * B = [1,2,3,4,5,6]
+ * print median(A) #should be 39
+ * print median(B) #should be 4
+ *
+ * @author [Benjamin Weiss](https://github.com/weiss-ben)
+ * @see median_search.cpp
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+
+/**
+ * Definition for singly-linked list.
+ */
+struct ListNode {
+    int val{0};               ///< the value stored in the node
+    ListNode* next{nullptr};  ///< pointer to the next node
+    ListNode() = default;     ///< default constructor
+    explicit ListNode(int x)
+        : val(x) {}  ///< constructor with value for node->val provided
+    ListNode(int x, ListNode* next)
+        : val(x),
+          next(next) {
+    }  ///< constructor with values provided for node->val and node->next
+};
+
+/**
+ * @namespace search
+ * @brief Search algorithms
+ */
+namespace search {
+/**
+ * @namespace median_search
+ * @brief Functions for the Median Search algorithm implementation. Wkipedia
+ * link to algorithm: https://en.wikipedia.org/wiki/Median_search
+ */
+namespace median_search2 {
+/**
+ * This function searches for the median of a linked list.
+ * @param head The head of the linked list.
+ * @returns Median node of the linked list.
+ */
+ListNode* middleNode(ListNode* head) {
+    if (!head) {
+        return nullptr;
+    }
+
+    // Fast and slow pointers
+    ListNode* fastptr = nullptr;
+    ListNode* slowptr = fastptr = head;
+
+    // fast jumps 2 while slow jumps 1
+    while (fastptr->next && fastptr->next->next) {
+        slowptr = slowptr->next;
+        fastptr = fastptr->next->next;
+    }
+    return (fastptr->next) ? slowptr->next : slowptr;
+}
+}  // namespace median_search2
+}  // namespace search
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    auto* head = new ListNode;
+    head->val = 1;
+
+    ListNode* temp1 = head;
+    for (int i = 1; i < 6; ++i) {
+        auto temp2 = new ListNode;
+        temp2->val = i;
+
+        temp1->next = temp2;
+        temp1 = temp2;
+    }
+
+    ListNode* median = search::median_search2::middleNode(head);
+    assert(3 == median->val);  // 3 is the value of the median node.
+    std::cout << "test case:1 passed\n";
+
+    // Clean up
+    while (head) {
+        ListNode* t = head;
+        head = head->next;
+        delete t;
+    }
+
+    head = new ListNode;
+    head->val = 1;
+
+    temp1 = head;
+    for (int i = 1; i < 7; ++i) {
+        auto temp2 = new ListNode;
+        temp2->val = i;
+
+        temp1->next = temp2;
+        temp1 = temp2;
+    }
+
+    median = search::median_search2::middleNode(head);
+    assert(4 == median->val);  // 3 is the value of the median node.
```

Comment:
Strangely, this doesn't seem to work when compiling on Gitpod.

![image](https://user-images.githubusercontent.com/51391473/192041220-9b30621c-15c7-4d30-ac41-7608b6675073.png)

Reply:
I guess it wouldn't be C++ if a wild pointer bug never appeared lol. I believe I have fixed it, it compiles locally and passes the tests.